Process timezone

dannyroberts opened the pull request:
The goal of this PR is to make it possible for a new project to use new timezone behavior—that should all work in terms of submissions being handled and reports showing correctly.

What this PR does not do:
- does not handle the migration for current domains
- does not make new timezones the default yet

dannyroberts committed:
process timezones in submissions when flag is set

dannyroberts committed:
fix couchforms test under new timezone behavior

dannyroberts committed:
split timezone toggles into started and completed

dannyroberts committed:
add a way to force tests to use post-tz migration behavior

dannyroberts committed:
fix timezone conversion tests

dannyroberts committed:
make AdjustDatetimesTest pass pre- and post- migration

dannyroberts committed:
run certain tests twice w/ pre & post tz migration behavior

dannyroberts committed:
have travis run tests using new migration behavior by default

dannyroberts committed:
temporarily have travis not fail fast

dimagimon commented:
There are 14 errors:

* corehq/apps/fixtures/models.py, line 7 - E501 line too long (162 > 115 characters)
* corehq/apps/userreports/reports/specs.py, line 2 - E501 line too long (121 > 115 characters)
* corehq/couchapps/models.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/corehq/couchapps/models.py'
* corehq/ex-submodules/couchforms/tests/test_adjust_datetimes.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/corehq/ex-submodules/couchforms/tests/test_adjust_datetimes.py'
* corehq/ex-submodules/phonelog/models.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/corehq/ex-submodules/phonelog/models.py'
* corehq/pillows/dynamic.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/corehq/pillows/dynamic.py'
* corehq/util/timezones/conversions.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/corehq/util/timezones/conversions.py'
* corehq/util/timezones/test_utils.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/corehq/util/timezones/test_utils.py'
* custom/_legacy/pathfinder/models.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/custom/_legacy/pathfinder/models.py'
* custom/apps/wisepill/models.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/custom/apps/wisepill/models.py'
* custom/ewsghana/models.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/custom/ewsghana/models.py'
* custom/fri/models.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/custom/fri/models.py'
* custom/m4change/reports/mcct_project_review.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/custom/m4change/reports/mcct_project_review.py'
* custom/uth/models.py, line 1 - E902 IOError: [Errno 2] No such file or directory: '/var/lib/lintreview/lint-review/workspace/dimagi/commcare-hq/6341/custom/uth/models.py'

dannyroberts committed:
fix pact test

dannyroberts committed:
make it easy to start using new tz behavior on new domains

czue commented:
put a "don't merge" tag on this. assume that's right

dannyroberts commented:
Yes! Thanks

dannyroberts committed:
fix pact test

to reproduce this test failure, I have to switch my local machine time to UTC.
There is likely an actual bug that will actually be triggered when migrating
but I am not sure it is worth fixing

dannyroberts committed:
add subtle cue that a domain has new timezone behavior enabled

dannyroberts commented:
@czue @snopoke right now in this PR a domain's timezone behavior (and very loosely their progress through an imagined future migration) is tracked with toggles. I'm wondering what you think about that. In my mind, the pros are:
- Was easy to configure
- comes with a UI for checking their values
- comes with utilities for setting and unsetting, checking values etc.
and cons are:
- it's not really what toggles are for; you should never be setting these manually because changing a domain's behavior requires a whole migration to take place, and there's no way to change it back once you've done that
- checking a few toggles on every submission is possibly slow (?)
- for the reality of getting the migration applied eventually to all domains, it doesn't make sense to store that information in two toggles, each of which are in a single couch document that has `enabled_users` set to the full list of domains on production.
- The migration may want to be able to store extra information—such has how far along it is, whether there were errors, and whatever else we at the last minute realize we need to add to the migration—and none of this can effectively be stored in toggles.

To me the alternative is to make a table in postgres that tracks this information for each domain, and to work off that. If I'm right in thinking how much of a pain using toggles is going to be a little down the road, it seems totally worth making that switch up front. Let me know whether you think I'm over thinking this

snopoke commented:
On the surface toggles seem the right choice but after reading through your
list of pros / cons I agree that it's going to become a headache down the
line and SQL table should make sense.

czue commented:
SQL for sure.

As you noted, a major downside of toggles is that the list of enabled XXX is all stored in a single couch doc, so if there are thousands of XXX that doc can get big and I would expect the overhead of pulling it becomes real (fortunately there's a decent amount of caching stuff built in that makes it not so painful).

It's been in the back of my mind that toggles also likely aren't the right answer for feature previews. Definitely out of scope, but if you think the thing you end up building for this would be useful for those would be good to have that requirement in the back of your head so we could roll it out later on.

TylerSheffels committed:
Merge remote-tracking branch 'origin/master' into process-timezone

Conflicts:
	corehq/toggles.py

dannyroberts committed:
Merge branch 'master' into process-timezone

dannyroberts committed:
Merge branch 'fix-circular-import' into process-timezone

dannyroberts committed:
track tz migration in a table rather than toggle

dimagimon commented:
There are 6 errors:

* corehq/apps/tzmigration/__init__.py, line 1 - F401 'get_migration_complete' imported but unused
* corehq/apps/tzmigration/__init__.py, line 1 - F401 'get_migration_status' imported but unused
* corehq/apps/tzmigration/__init__.py, line 1 - F401 'phone_timezones_have_been_processed' imported but unused
* corehq/apps/tzmigration/__init__.py, line 1 - F401 'phone_timezones_should_be_processed' imported but unused
* corehq/apps/tzmigration/__init__.py, line 1 - F401 'set_migration_complete' imported but unused
* corehq/apps/tzmigration/api.py, line 3 - F401 'quickcache' imported but unused

snopoke commented:
Test fail real:

```
Traceback (most recent call last):
  File "manage.py", line 65, in <module>
    execute_from_command_line(sys.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 50, in run_from_argv
    super(Command, self).run_from_argv(argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 242, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 71, in execute
    super(Command, self).execute(*args, **options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/management/commands/test.py", line 8, in handle
    super(Command, self).handle(*args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 88, in handle
    failures = test_runner.run_tests(test_labels)
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 284, in run_tests
    all_suite = self.build_suite(all_test_labels, extra_tests)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/simple.py", line 244, in build_suite
    suite.addTest(build_suite(app))
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/simple.py", line 149, in build_suite
    test_module = get_tests(app_module)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/simple.py", line 101, in get_tests
    test_module = import_module('.'.join(prefix + [TEST_MODULE]))
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/importlib.py", line 40, in import_module
    __import__(name)
  File "corehq/ex-submodules/casexml/apps/case/tests/__init__.py", line 5, in <module>
    from casexml.apps.case.tests.util import delete_all_cases, delete_all_xforms
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 15, in <module>
    from couchforms.tests.testutils import post_xform_to_couch
  File "corehq/ex-submodules/couchforms/tests/__init__.py", line 20, in <module>
    raise(e)
ImportError: No module named test_utils
```

snopoke commented on corehq/apps/hqwebapp/templates/hqwebapp/partials/global_navigation_bar.html:
```diff
@@ -31,7 +32,13 @@
+                    <i class="icon-globe" title="Timezone migration started but not complete"></i>
+                    {% else %}
                     <i class="icon-cog"></i>
+                    {% endif %}
```
Should this also be added to the [BS3 version](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/style/templates/style/includes/navigation_global.html#L33)?

dannyroberts replied:
Hmm, well that's a good point, though I think it's already different (e.g. I don't see the cog icon that is in this file); and maybe more to the point, this is sort of a short-term thing to make it easy for me to quickly distinguish when looking at a domain whether it's been migrated. (Like when I'm debugging.) I'll probably change it and/or remove it eventually and I think right now bs3 isn't used anywhere except dashboard, so will be nice to not have to try and keep those in sync. Let me know if that's too lazy of me and I'll update bs3 template too.

snopoke replied:
Good argument, fine with me

snopoke commented on corehq/apps/tzmigration/api.py:
```diff
@@ -0,0 +1,81 @@
+        override = getattr(
+            settings, 'PHONE_TIMEZONES_HAVE_BEEN_PROCESSED', None)
+        if override is not None:
+            return override
```
I like this pattern for testing. Much easier than mocking.

dannyroberts replied:
Yeah it seems to have worked well in this case.

snopoke commented on corehq/apps/tzmigration/api.py:
```diff
@@ -0,0 +1,81 @@
+        MigrationStatus.IN_PROGRESS, MigrationStatus.COMPLETE)
+
+
+def _get_migration_status_from_threadlocals():
```
I can see a few other places we can use this. e.g. callcenter, user as a case

dannyroberts replied:
"this" being pulling a flag based on the thread-local request? I'm finding to be a nice pattern, it seems to work robustly, and no one on the internet has told me it's a bad idea...

dannyroberts replied:
Obvious downside is it relies on a function being processed in the same thread as the triggering request was made, which makes it harder to refactor something to be done asynchronously (with e.g. celery) or by a separate little service if/when we start going that route.

snopoke replied:
Yup, that. I assume it works with gevent patched threads?

dannyroberts replied:
Yeah for sure. I think gevent patches it to work on gevent "threads".

snopoke commented on corehq/apps/tzmigration/models.py:
```diff
@@ -0,0 +1,34 @@
+                              db_index=True, primary_key=True)
+    migration_status = models.CharField(choices=MigrationStatus.choices,
+                                        max_length=MigrationStatus.max_length,
+                                        default=MigrationStatus.NOT_STARTED)
```
Just wondering about making this generic. What do you think of this:

```python
class SqlToggle(models.Model):
    default_name = None
    values = {
        'disabled': 0,
        'enabled': 1
    }

    domain = models.TextField(null=False, blank=False, db_index=True)
    name = models.TextField(null=False, blank=False, db_index=True)
    value = models.IntegerField(null=True)

    def save(self, *args, **kwargs):
        if not self.name and self.default_name:
            self.name = self.default_name

        super(SqlToggle, self).save(*args, **kwargs)

    def __getattr__(self, item):
        if item.startswith('is_') and item[3:] in self.values:
            expected = self.values[item[3:]]
            return self.value == expected
        super(SqlToggle, self).__getattr__(item)

    class Meta:
        unique_together = ('domain', 'name')


class TimezoneMigrationProgress(SqlToggle):
    default_name = 'TimezoneMigrationProgress'
    values = {
        'not_started': 0,
        'in_progress': 1,
        'complete': 2
    }

    class Meta:
        proxy = True

TimezoneMigrationProgress.get_or_create(domain='a').is_in_progress
```

dannyroberts replied:
Yeah, I'm just not really sure in this case because I may also want to record progress through the migration (number of forms migrated, number of cases migrated, maybe a checkpoint of some sort depending on how we do the migration, etc.). Anyways, I didn't want to engineer a general solution for a single case when I'm not actually totally sure how-all I'm going to use it.

snopoke replied:
Makes sense

snopoke commented on corehq/apps/tzmigration/test_utils.py:
```diff
@@ -0,0 +1,52 @@
+                raise
+
+
+def run_with_multiple_settings(fn, settings_dicts):
```
Nice

snopoke commented on corehq/toggles.py:
```diff
@@ -395,16 +395,13 @@ def toggles_dict(username=None, domain=None):
+BULK_PAYMENTS = StaticToggle(
+    'bulk_payments',
+    'Enable payment of invoices by bulk credit payments and invoice generation for wire transfers',
+    TAG_PRODUCT_CORE
```
I assume this is from a merge?

dannyroberts replied:
Must be. I think there was a conflict. Might be worth rebasing as a way to make sure

snopoke commented:
Looking good @dannyroberts what's outstanding / next steps?

dannyroberts committed:
Merge branch 'master' into process-timezone

dannyroberts committed:
fix import

dannyroberts committed:
Revert "temporarily have travis not fail fast"

This reverts commit 79b3800f157422eac31cdc1addde42a0c4f484db.

dannyroberts commented:
Well, to some extent this is fairly innocuous to merge because ~everything is under a flag. (Of course as always, there's some possibility that the flag logic itself could be messed up, etc.) In addition I've tested on staging that a domain with the flag set to migration complete processes forms "correctly" the new way. In the process of testing I noticed we made no attempt to think about timezones in submit history (!) so I made this PR: https://github.com/dimagi/commcare-hq/pull/6827, but that's only conceptually related, not directly blocking.

dannyroberts commented:
So I'd say next steps are:
- wait for tests to pass of course
- I'll put this back on staging and poke around and submit a form to a domain with the flag set to not_started and complete. (I'm not going to test in_progress just yet because I think what I'm eventually going to do is just 503 or something on submissions to domains that with a migration in progress, to avoid having to deal with the edge cases.)

@snopoke @czue Anyone want to push

czue commented:
I haven't reviewed closely but happy to defer to you and @snopoke's judgment on QA.

snopoke commented:
Merged, I think the risk is fairly low.

dannyroberts commented:
Woohoo! This is an important milestone for what's been a long and slow process! Next steps:
- Little thing, https://github.com/dimagi/commcare-hq/pull/6849 for getting timeEnd etc in the right timezone on form data page (both pre and post migration)
- QA a single new domain that's manually marked as migrated in production. Make sure that the new semantics are all correct.
- Make good tz behavior the default for new domains
- In parallel, write and test the code to migrate an existing domain
- start migrating domains, starting with small ones (and here's where the fun begins...)
